beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,264 @@
+   */
+  double min_particle_weight{0.08};
+
+  /// Dataset path.
+  /**
+   * The dataset file is used to save the data produced by the simulation for posterior analisys.
+   */
+  std::filesystem::path record_path{"./record.yaml"};
+};
+
+using Particle = std::tuple<double, beluga::Weight>;
+using LandmarkMapVector = std::vector<double>;
+using SensorData = std::vector<double>;
+using GroundTruth = double;
+using Estimate = std::pair<double, double>;
+
+struct RobotRecord {
+  GroundTruth ground_truth;
+  beluga::TupleVector<Particle> current;
+  beluga::TupleVector<Particle> propagate;
+  beluga::TupleVector<Particle> reweight;
+  beluga::TupleVector<Particle> resample;
+  Estimate estimation;
+};
+
+struct SimulationRecord {
+  LandmarkMapVector landmark_map;
+  std::vector<RobotRecord> robot_record;
+};
+
+void load_parameters(const std::filesystem::path& path, Parameters& parameters) {
+  try {
+    YAML::Node params = YAML::LoadFile(path);
+    parameters.map_size = params["map_size"].as<int>();
+    parameters.number_of_doors = params["number_of_doors"].as<int>();
+    parameters.number_of_particles = params["number_of_particles"].as<int>();
+    parameters.number_of_cycles = params["number_of_cycles"].as<unsigned int>();
+    parameters.initial_position = params["initial_position"].as<double>();
+    parameters.initial_position_sd = params["initial_position_sd"].as<double>();
+    parameters.dt = params["dt"].as<double>();
+    parameters.velocity = params["velocity"].as<double>();
+    parameters.translation_sd = params["translation_sd"].as<double>();
+    parameters.sensor_range = params["sensor_range"].as<double>();
+    parameters.sensor_model_sigma = params["sensor_model_sigma"].as<double>();
+    parameters.min_particle_weight = params["min_particle_weight"].as<double>();
+    parameters.record_path = params["record_path"].as<std::string>();
+  } catch (YAML::BadFile& e) {
+    std::cout << e.what() << "\n";
+  }
+}
+
+void load_landmark_map(const std::filesystem::path& path, LandmarkMapVector& landmark_map) {
+  try {
+    YAML::Node node = YAML::LoadFile(path);
+    landmark_map = node["landmark_map"].as<LandmarkMapVector>();
+  } catch (YAML::BadFile& e) {
+    std::cout << e.what() << "\n";
+  }
+}
+
+void save_simulation_records(const std::filesystem::path& path, const SimulationRecord& simulation_record) {
+  YAML::Node node;
+
+  for (const auto& landmark : simulation_record.landmark_map) {
+    node["landmark_map"].push_back(landmark);
+  }
+
+  for (auto&& [cycle, record] : simulation_record.robot_record | ranges::views::enumerate) {
+    node["simulation_records"][cycle]["ground_truth"] = record.ground_truth;
+
+    auto current = node["simulation_records"][cycle]["particles"]["current"];
+    current["states"] = beluga::views::states(record.current) | ranges::to<std::vector<double>>;
+    current["weights"] = beluga::views::weights(record.current) | ranges::to<std::vector<double>>;
```

Comment:
meta: I feel like there is an opportunity for reducing code duplication by specializing [`YAML::convert`](https://github.com/jbeder/yaml-cpp/blob/master/include/yaml-cpp/node/convert.h) for particle range types. Also, it would be nice to have sensor, motion, and particle filter parameters separated in their own structures.

No need to address this now, I can give it a stab in a follow-up PR.